### PR TITLE
DAOS-11630 test: Add md on ssd replay test case

### PR DESCRIPTION
**Pull request by phender**
Add test case to verify data access after engine restart with WAL replay and check pointing with unsynchronized WAL and VOS.

Skip-unit-tests: true
Test-tag: test_restart
Test-nvme: auto_md_on_ssd

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/tests/ftest/server/replay.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
+"""
+  (C) Copyright 2019-2023 Intel Corporation.
```
Comment (daltonbohning) [inline src/tests/ftest/server/replay.py:2]: Minor - Should this be?
```suggestion
  (C) Copyright 2023 Intel Corporation.
```

File: src/tests/ftest/server/replay.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+"""
+
+from apricot import TestWithServers
+from ior_utils import run_ior
+from job_manager_utils import get_job_manager
+from test_utils_pool import add_pool
+
+
+class ReplayTests(TestWithServers):
```
Comment (daltonbohning) [inline src/tests/ftest/server/replay.py:13]: I don't think we have a formal pattern, but how do you feel about putting the directory name in the class name? E.g.
```suggestion
class ServerReplayTest(TestWithServers):
```

File: src/tests/ftest/server/replay.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    """Shutdown/restart/replay test cases.
+
+    Restarting engines with volatile SCM will include loading the blob from the SSD and re-applying
+    any changes from the WAL.
+
+    :avocado: recursive
+    """
+
+    def test_restart(self):
```
Comment (daltonbohning) [inline src/tests/ftest/server/replay.py:22]: Similar to previous comment, how do we ensure these are unique? E.g. maybe?
```suggestion
    def test_server_replay_restart(self):
```

File: src/tests/ftest/server/replay.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+            4) Remove VOS file manually/temporarily (umount tmpfs; remount tmpfs)
+            5) Restart each engine (dmg system start)
+            6) Verify the previously written data matches with an ior read
+
+        :avocado: tags=all,pr
+        :avocado: tags=hw,medium
+        :avocado: tags=server,replay
+        :avocado: tags=ReplayTests,test_restart
+        """
+        processes = self.params.get('/run/ior_write/*', 'ppn', 1)
```
Comment (daltonbohning) [inline src/tests/ftest/server/replay.py:41]: Typo, or am I misunderstanding?
```suggestion
        processes = self.params.get('ppn', '/run/ior_write/*', 1)
```

File: src/tests/ftest/server/replay.yaml [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+server_config:
+  name: daos_server
+  engines_per_host: 1
+  engines:
+    0:
+      storage: auto
+
+pool:
+  size: 50%
+  control_method: dmg
```
Comment (daltonbohning) [inline src/tests/ftest/server/replay.yaml:16]: FYI - not used
```suggestion
```

File: src/tests/ftest/server/replay.yaml [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+  control_method: daos
+
+ior: &ior_base
+  client_processes:
+    ppn: 4
+  api: DFS
+  transfer_size: 512K
+  block_size: 1G
+  dfs_oclass: SX
+  dfs_dir_oclass: SX
```
Comment (daltonbohning) [inline src/tests/ftest/server/replay.yaml:28]: FYI - IOR doesn't use dfs_dir_oclass. But we specify it all over
```suggestion
```

**Review by daltonbohning — APPROVED** [review]
Just nit picks and suggestions

File: src/tests/ftest/server/replay.py (hunk 2 above)
Comment (daltonbohning) [reply src/tests/ftest/server/replay.py:13]: Yeah, I don't disagree with that. I'm thinking more for the sake of guaranteeing unique test method names, but that's a topic for another time

File: src/tests/ftest/server/replay.yaml (hunk 5 above)
Comment (phender) [reply src/tests/ftest/server/replay.yaml:16]: Fixed

File: src/tests/ftest/server/replay.yaml (hunk 6 above)
Comment (phender) [reply src/tests/ftest/server/replay.yaml:28]: Fixed

File: src/tests/ftest/server/replay.py (hunk 4 above)
Comment (phender) [reply src/tests/ftest/server/replay.py:41]: Fixed

**Review by daltonbohning — APPROVED** [review]

File: src/tests/ftest/util/ior_utils.py [hunk 7]
```diff
@@ -467,6 +486,18 @@ def command(self):
...
+
+        Args:
+            name (str): name of the IorCommand BasicParameter to update
+            value (str): value to assign to the IorCommand BasicParameter
+        """
+        param = getattr(self.command, name, None)
+        if param:
+            if isinstance(param, BasicParameter):
+                param.update(value, ".".join([self.command.command, name]))
+
```
Comment (daltonbohning) [inline src/tests/ftest/util/ior_utils.py:500]: Suggestion - how about defining `update_params` similar to `ObjectWithParameters.update_params`?

**Review by daltonbohning — APPROVED** [review]

**Comment by phender** [comment]
@daos-stack/daos-gatekeeper the updated ticket number for this PR is DAOS-13009.  Please use this ticket when squashing comments.

**Review by rpadma2 — APPROVED** [review]

**Pull request merged**